src/runtime/math.py

```diff
@@ -15,36 +15,169 @@ def modf(x: f64) -> tuple[f64, f64]:
     """
     return (x - f64(int(x)), float(int(x)))
 
-@overload
-def factorial(x: i32) -> i32:
-    """
-    Computes the factorial of `x`.
-    """
 
-    result: i32
-    result = 0
-    if x < 0:
-        return result
-    result = 1
-    i: i32
-    for i in range(1, x+1):
-        result *= i
-    return result
+def factorial(n: i32) -> i64:
+    """Computes the factorial of `n`."""
+    MAX_LOOKUP_VALUE: i32 = 20
+    FACTORIAL_LOOKUP_TABLE: list[i64] = [
+        i64(1),
+        i64(1),
+        i64(2),
+        i64(6),
+        i64(24),
+        i64(120),
+        i64(720),
+        i64(5040),
+        i64(40320),
+        i64(362880),
+        i64(3628800),
+        i64(39916800),
+        i64(479001600),
+        i64(6227020800),
+        i64(87178291200),
+        i64(1307674368000),
+        i64(20922789888000),
+        i64(355687428096000),
+        i64(6402373705728000),
+        i64(121645100408832000),
+        i64(2432902008176640000),
+    ]
+    if n < 0:
+        # Exceptions are not implemented currently
+        # raise ValueError("factorial() not defined for negative values")
+        assert 1 == 0, "factorial() not defined for negative values."
+    elif n < MAX_LOOKUP_VALUE:
+        return FACTORIAL_LOOKUP_TABLE[n]
+    else:
+        f: list[i32] = [0] * 4300
+        f[0] = 0
+        f[1] = 0
+        f[2] = 0
+        f[3] = 0
+        f[4] = 4
+        f[5] = 6
+        f[6] = 6
+        f[7] = 7
+        f[8] = 1
+        f[9] = 8
+        f[10] = 0
+        f[11] = 0
+        f[12] = 2
+        f[13] = 0
+        f[14] = 9
+        f[15] = 2
+        f[16] = 3
+        f[17] = 4
+        f[18] = 2
+
+        f_size: i32 = 19
+
+        i: i32 = 21
+        while i <= n:
+            index: i32 = 0
+            carry: i32 = 0
+            while index < f_size:
+                product: i32 = f[index] * i + carry
+                f[index] = product % 10
+
+                carry = product // 10
+                index += 1
+
+            while carry > 0:
+                f[f_size] = carry % 10
+                carry = carry // 10
+                f_size += 1
+            i += 1
+
+    result: str = ""
+    idx: i32
+    for idx in range(f_size - 1, -1, -1):
+        result += str(f[idx])
+    print(result)
+    return i64(0)
```

Comment:
I do understand the approach, here the final result is printed out using concatenation.
I have not looked into the pre-requisites but even after a conversion from string to int, how will you return an integer here? The highest supported annotation we have is `i64`, which has a maximum value of `2^63 - 1`. 
This is not large enough to store anything over `20!`, which means this will still encounter an overflow and return a garbage value.

Reply:
I have tried something similar in another project, let me know if there are any other approaches. Thanks :)

Reply:
@faze-geek I understand your point. Internally in LPython, the method used for handling values larger than 2 ^ 63 - 1 is handled by the BigInt module we have. The part of the module which does the handling for large numbers is broken and returns a pointer to the value instead. I propose implementing the vector and string storage type for handling large values. Only then will we be able to handle values like `20!`.

I had forgotten to mention this prerequisite. Thanks for reminding me. I am adding it above.

Reply:
> I have tried something similar in another project, let me know if there are any other approaches. Thanks :)

Oh yes there are! As I had mentioned above, the industry standard for scientific computing applications is to use the [PrimeSwing algorithm](https://www.luschny.de/math/factorial/SwingFactorialPy.html). The Python implementation is available online, but I do not want to type something I do not understand. :smile: 

The Python `math.factorial(n)` uses (this)[http://www.luschny.de/math/factorial/binarysplitfact.html] algorithm.

Reply:
Thanks. Will surely explore this once the `BigInt` module gets working.